extractor/filesystem/containers/k8simage/k8simage_test.go

```diff
@@ -0,0 +1,286 @@
+					PURLType:  purl.TypeDocker,
+				},
+				{
+					Name:      "nginx",
+					Version:   "1.7.9",
+					Locations: []string{"testdata/deployment.yaml"},
+					PURLType:  purl.TypeDocker,
+				},
+				{
+					Name:      "my-demo-app",
+					Version:   "latest",
+					Locations: []string{"testdata/deployment.yaml"},
+					PURLType:  purl.TypeDocker,
+				},
+				{
+					Name:      "alpine",
+					Version:   "3.7",
+					Locations: []string{"testdata/deployment.yaml"},
+					PURLType:  purl.TypeDocker,
+				},
+			},
+		},
+		{
+			name: "pod with multiple documents",
+			path: "testdata/pod.yaml",
+			cfg:  k8simage.DefaultConfig(),
+			wantPackages: []*extractor.Package{
+				{
+					Name:      "gcr.io/google-samples/hello-app",
+					Version:   "2.0",
+					Locations: []string{"testdata/pod.yaml"},
+					PURLType:  purl.TypeDocker,
+				},
+				{
+					Name:      "gcr.io/google-samples/hello-app",
+					Version:   "1.0",
+					Locations: []string{"testdata/pod.yaml"},
+					PURLType:  purl.TypeDocker,
+				},
+			},
+		},
+		{
+			name: "job with single container",
+			path: "testdata/job.yaml",
+			cfg:  k8simage.DefaultConfig(),
+			wantPackages: []*extractor.Package{
+				{
+					Name:      "perl",
+					Version:   "5.34.0",
+					Locations: []string{"testdata/job.yaml"},
+					PURLType:  purl.TypeDocker,
+				},
+			},
+		},
+		{
+			name: "replicaset with single container",
+			path: "testdata/frontend.yaml",
+			cfg:  k8simage.DefaultConfig(),
+			wantPackages: []*extractor.Package{
+				{
+					Name:      "us-docker.pkg.dev/google-samples/containers/gke/gb-frontend",
+					Version:   "v5",
+					Locations: []string{"testdata/frontend.yaml"},
+					PURLType:  purl.TypeDocker,
+				},
+			},
+		},
+		{
+			name: "multi-document with StatefulSet and Service",
+			path: "testdata/service.yaml",
+			cfg:  k8simage.DefaultConfig(),
+			wantPackages: []*extractor.Package{
+				{
+					Name:      "registry.k8s.io/kubernetes-zookeeper",
+					Version:   "1.0-3.4.10",
+					Locations: []string{"testdata/service.yaml"},
+					PURLType:  purl.TypeDocker,
+				},
+			},
+		},
+		{
+			name: "StatefulSet with single container",
+			path: "testdata/StatefulSet.yaml",
+			cfg:  k8simage.DefaultConfig(),
+			wantPackages: []*extractor.Package{
+				{
+					Name:      "gcr.io/google-samples/cassandra",
+					Version:   "v13",
+					Locations: []string{"testdata/StatefulSet.yaml"},
+					PURLType:  purl.TypeDocker,
+				},
+			},
+		},
+		{
+			name: "multi-document with Service and StatefulSet",
+			path: "testdata/web.yaml",
+			cfg:  k8simage.DefaultConfig(),
+			wantPackages: []*extractor.Package{
+				{
+					Name:      "registry.k8s.io/nginx-slim",
+					Version:   "0.21",
+					Locations: []string{"testdata/web.yaml"},
+					PURLType:  purl.TypeDocker,
+				},
+				{
+					Name:      "registry.k8s.io/nginx2-slim",
+					Version:   "0.21",
+					Locations: []string{"testdata/web.yaml"},
+					PURLType:  purl.TypeDocker,
+				},
+			},
+		},
+		{
```

Comment:
How about a testcase with `containers` containing malformed or missing sub elements? Or would this lead to parsing error?